src/ray/gcs/gcs_client/service_based_accessor.cc

```diff
@@ -549,22 +597,67 @@ Status ServiceBasedNodeInfoAccessor::AsyncSubscribeBatchHeartbeat(
     const StatusCallback &done) {
   RAY_LOG(DEBUG) << "Subscribing batch heartbeat.";
   RAY_CHECK(subscribe != nullptr);
-  auto on_subscribe = [subscribe](const ClientID &node_id,
-                                  const HeartbeatBatchTableData &data) {
-    subscribe(data);
+  auto on_subscribe = [subscribe](const std::string &id, const std::string &data) {
+    HeartbeatBatchTableData heartbeat_batch;
+    heartbeat_batch.ParseFromString(data);
+    subscribe(heartbeat_batch);
   };
-  auto status = heartbeat_batch_sub_executor_.AsyncSubscribeAll(ClientID::Nil(),
-                                                                on_subscribe, done);
+  auto status = gcs_sub_.SubscribeAll(heartbeat_batch_channel_, on_subscribe, done);
   RAY_LOG(DEBUG) << "Finished subscribing batch heartbeat.";
   return status;
 }
 
+void ServiceBasedNodeInfoAccessor::HandleNotification(const GcsNodeInfo &node_info) {
+  ClientID node_id = ClientID::FromBinary(node_info.node_id());
+  bool is_alive = (node_info.state() == GcsNodeInfo::ALIVE);
+  // It's possible to get duplicate notifications from the client table, so
+  // check whether this notification is new.
+  auto entry = node_cache_.find(node_id);
+  bool is_notif_new;
+  if (entry == node_cache_.end()) {
+    // If the entry is not in the cache, then the notification is new.
+    is_notif_new = true;
+  } else {
+    // If the entry is in the cache, then the notification is new if the client
+    // was alive and is now dead or resources have been updated.
+    bool was_alive = (entry->second.state() == GcsNodeInfo::ALIVE);
+    is_notif_new = was_alive && !is_alive;
+    // Once a client with a given ID has been removed, it should never be added
+    // again. If the entry was in the cache and the client was deleted, check
+    // that this new notification is not an insertion.
+    if (!was_alive) {
+      RAY_CHECK(!is_alive)
+          << "Notification for addition of a client that was already removed:" << node_id;
+    }
+  }
+
+  // Add the notification to our cache. Notifications are idempotent.
+  RAY_LOG(INFO) << "[ClientTableNotification] ClientTable Insertion/Deletion "
+                   "notification for client id "
+                << node_id << ". IsAlive: " << is_alive
+                << ". Setting the client cache to data.";
```

Comment:
This seems to be copy-pasted from the old `table.cc` file. We should update this, as we don't have "ClientTable" now.

Reply:
So are other comments that use the term "client".

Reply:
fixed